curve25519-dalek/src/montgomery.rs

```diff
@@ -252,33 +254,98 @@ impl MontgomeryPoint {
-    let d_1 = &one + &r_0.square2(); /* 2r^2 */
-
-    let d = &MONTGOMERY_A_NEG * &(d_1.invert()); /* A/(1+2r^2) */
-
-    let d_sq = &d.square();
-    let au = &MONTGOMERY_A * &d;
-
-    let inner = &(d_sq + &au) + &one;
-    let eps = &d * &inner; /* eps = d^3 + Ad^2 + d */
-
-    let (eps_is_sq, _eps) = FieldElement::sqrt_ratio_i(&eps, &one);
-
-    let zero = FieldElement::ZERO;
-    let Atemp = FieldElement::conditional_select(&MONTGOMERY_A, &zero, eps_is_sq); /* 0, or A if nonsquare*/
-    let mut u = &d + &Atemp; /* d, or d+A if nonsquare */
-    u.conditional_negate(!eps_is_sq); /* d, or -d-A if nonsquare */
-
-    (MontgomeryPoint(u.to_bytes()), eps_is_sq)
+    let two = &one + &one;
+    let c2 = &two * &(two.pow_p58());
+
+    // 1.  tv1 = u^2
+    // 2.  tv1 = 2 * tv1
+    let tv1 = u.square2();
+    // 3.   xd = tv1 + 1
+    let xd = &one + &tv1;
+    // 4.  x1n = -J
+    let x1n = MONTGOMERY_A_NEG;
+    // 5.  tv2 = xd^2
+    let tv2 = xd.square();
+    // 6.  gxd = tv2 * xd
+    let gxd = &tv2 * &xd;
+    // 7.  gx1 = J * tv1
+    let gx1 = &MONTGOMERY_A * &tv1;
+    // 8.  gx1 = gx1 * x1n
+    let gx1 = &gx1 * &x1n;
+    // 9.  gx1 = gx1 + tv2
+    let gx1 = &gx1 + &tv2;
+    // 10. gx1 = gx1 * x1n
+    let gx1 = &gx1 * &x1n;
+    // 11. tv3 = gxd^2
+    let tv3 = gxd.square();
+    // 12. tv2 = tv3^2
+    let tv2 = tv3.square();
+    // 13. tv3 = tv3 * gxd
+    let tv3 = &tv3 * &gxd;
+    // 14. tv3 = tv3 * gx1
+    let tv3 = &tv3 * &gx1;
+    // 15. tv2 = tv2 * tv3
+    let tv2 = &tv2 * &tv3;
+    // 16. y11 = tv2^c4
+    let y11 = tv2.pow_p58();
+    // 17. y11 = y11 * tv3
+    let y11 = &y11 * &tv3;
+    // 18. y12 = y11 * c3
+    let y12 = &y11 * &SQRT_M1;
+    // 19. tv2 = y11^2
+    let tv2 = y11.square();
+    // 20. tv2 = tv2 * gxd
+    let tv2 = &tv2 * &gxd;
+    // 21.  e1 = tv2 == gx1
+    let e1 = tv2.ct_eq(&gx1);
+    // 22.  y1 = CMOV(y12, y11, e1)
+    let y1 = FieldElement::conditional_select(&y12, &y11, e1);
+    // 23. x2n = x1n * tv1
+    let x2n = &x1n * &tv1;
+    // 24. y21 = y11 * u
+    let y21 = &y11 * u;
+    // 25. y21 = y21 * c2
+    let y21 = &y21 * &c2;
+    // 26. y22 = y21 * c3
+    let y22 = &y21 * &SQRT_M1;
+    // 27. gx2 = gx1 * tv1
+    let gx2 = &gx1 * &tv1;
+    // 28. tv2 = y21^2
+    let tv2 = y21.square();
+    // 29. tv2 = tv2 * gxd
+    let tv2 = &tv2 * &gxd;
+    // 30.  e2 = tv2 == gx2
+    let e2 = tv2.ct_eq(&gx2);
+    // 31.  y2 = CMOV(y22, y21, e2)
+    let y2 = FieldElement::conditional_select(&y22, &y21, e2);
+    // 32. tv2 = y1^2
+    let tv2 = y1.square();
+    // 33. tv2 = tv2 * gxd
+    let tv2 = &tv2 * &gxd;
+    // 34.  e3 = tv2 == gx1
+    let e3 = tv2.ct_eq(&gx1);
+    // 35.  xn = CMOV(x2n, x1n, e3)
+    let xn = FieldElement::conditional_select(&x2n, &x1n, e3);
+    // 36.   y = CMOV(y2, y1, e3)
+    let y = FieldElement::conditional_select(&y2, &y1, e3);
+    // 37.  e4 = sgn0(y) == 1
+    // TODO: check that this step is correct
+    let e4 = y.is_negative();
```

Comment:
This is indeed correct, see https://www.rfc-editor.org/rfc/rfc9380.html#section-4-4.5.
I didn't check the implementation of `is_negative()` though.

Reply:
Oh, leftover of notes to myself. I also checked, and I believe it is correct 👍  I'll remove the TODO.

Reply:
Fixed